_posts/2024-08-27-cheri-myths-safe-languages.markdown

```diff
@@ -0,0 +1,156 @@
+When I joined the CHERI project in 2012, my primary motivation was producing hardware that enabled safe interoperability between languages, so I never felt that safe languages and CHERI were in tension.
+Quite the reverse: My work on CHERI was motivated by a desire to enable safe language adoption.
+
+# Safety is a property of entire programs
+
+The [WannaCry ransomware attack](https://en.wikipedia.org/wiki/WannaCry_ransomware_attack), which cost billions of dollars, was enabled by a single use-after-free bug.
+The [CrowdStrike incident](https://en.wikipedia.org/wiki/CrowdStrike#2024_CrowdStrike_incident), which also cost billions of dollars, was the result of an uninitialised-use bug (compounded by a lot of operational failures).
+
+Memory-safety bugs are particularly bad because they step outside of the language's abstract machine.
+After a program passes a memory-safety bug, it may do anything.
+It may access objects that should not be reachable, modify immutable objects, and even execute instructions out of sequence to do things that are not part of the source-code description of what the program should do at all.
+
+A single memory-safety bug can be all that an attacker needs to get an arbitrary-code execution exploit.
+Rewriting 99% of a program in a safe language but leaving just one memory-safety bug does not prevent at-scale deployment of malware.
+
+This was one of the problems that CHERI aimed to solve, in two ways.
+At the fine granularity, CHERI offers language-agnostic memory safety primitives.
+Compilers can generate, or humans can write, assembler making use of these primitives to capture and then enforce memory safety of source-level objects.
+The switcher is the most privileged part of a CHERIoT system and is written in assembly (it has to do things like save all registers, which are not possible in any high-level language).
+In spite of its privilege, it remains bound by the memory-safety rules.
+It cannot access any compartments other than the caller and callee.
+It cannot access any memory other than that reachable from the registers when it is invoked and its internal state, unlike a traditional OS kernel that can access anything.
+
+Enforcing these guarantees in the hardware makes it possible for safe-language code to call unsafe or, most excitingly, _differently_ safe code and yet still enforce source-level properties.
+Consider a single program composed of code written in all of Rust, Java, and Haskell.
+Each is in the 'safe languages' category, but provides different guarantees.
+Rust code passing a borrowed reference to Java may wish to ensure that it isn't captured, a property that CHERIoT can enforce directly (or CHERI can enforce with indirection).
+Haskell passing an object to Java or Rust may wish to guarantee deep immutability, a property that CHERIoT and all Morello or later CHERI platforms can enforce.
+
```

Comment:
I think there should be something around here (or back a paragraph or two) to remind that CHERI catches would-be memory safety violations _before_ they happen, enabling recovery within the abstract machine's semantics.